### ISSUE TITLE
cross-validation

### ISSUE THREAD

**Issue by jasperdev** [RUG-IDS/Course-2017#26]
Hi,

The exercise states that we have to do N-fold for cross validation. However the scores produced by this scheme are hard to compare; classifier scores are fairly similar and have wide confidence intervals. e.g. it is hard to say that a classifier scoring 95% is any better than another scoring 90%. One part of the problem is that the number of folds are limited because of the low number of `sick` examples. 

Is it OK to use a more informative scheme like random permutation cross-validation?

**Comment by sreejita-rug** [comment]
Hi Jasper,
1) Since the dataset is imbalanced (majority healthy and minority sick/patients), hence in order to ensure that atleast some patients are present during training perform the split separately for for the classes. That is, first partition the labelled data into healthy and patients. Then do N fold cross-validation with any of the inbuilt data spilling schemes in R or Matlab or Python (whichever you are comfortable with, I had done it in Matlab). Do the data split separately for healthy and patients, and then for each fold put the selected indices for healthy and selected indices for patients together.  So now you have for example, 50 healthy and 5 patients (so 55 subjects) in fold 1, 48 healthy and 6 patients in fold 2 (54 subjects), so on for N folds. However, if you want you can also try random permutation cross validation to compare your results.
2) What exact measures are you referring to by 'classifier score'? Since you are dealing with an imbalanced class problem, be careful that overall accuracy is not enough, and can often be misleading. You might have 90% overall accuracy with all samples classified as healthy since healthy subjects are a majority. However this is unacceptable in the medical domain where the aim is to diagnose the sick ones. You should look at classwise accuracy, check the confusion matrices, receiver-operating-characteristics (ROC) curve, in order to determine if the classifier is really trained well. And I know atleast for MATLAB there are inbuilt functions for confusion matrices, and ROC curves. I googled and found that R also has a similar package for ROC curves. So obtaining these measures should be easy. Then check the results from the validation fold. For a particular classifier, suppose Knn with a certain setting, maybe you have comparable confusion matrices and class-wise accuracy for all your classifier models from 5 or 10 folds. However if the validation set performance is not comparable to the training set performance then there is overfitting. So you have to look at all these measures for both training and validation sets for all the models of your cross-validation. And when you see that for all the N folds a classifier has less overfitting, has a good class-wise accuracy, or area under the curve (AUC) of the ROC curve is nearly 1 (where True positive is the case AML/sick/patients) then you can select that as one of the candidates for your ensemble. 
Hope this is clear enough about how to partially deal with the limited number of sick, and how to decide which classifiers and which setting, and which model is better.

**Comment by sreejita-rug** [comment]
@jasperdev  plus during cross-validation scheme one would also like the models to be stable, so not much different from each other based on the training data the model in each fold gets. N-fold crossvalidation is performed so that the classifier is trained on different sets of the data and is not sensitive to minor differences. The best performances which you are to report does not imply which classifier model in which fold. Rather, it means which classifier with what kind of settings, entire datatset or with feature selection or reduction or if you did something to handle the imbalanced classes, and if a certain classifier with either one or any combination of these settings, or just the classifier w/o any such settings, produced a stable model (whose performance from one fold to another does not vary much). So if you have your classifier with class-wise accuracy of more than 90% for each of the 2 classes for each of your N folds, your classifier models seem to be stable
.

**Comment by jasperdev** [comment]
Thanks for your quick reply.

1. I was already using a stratified variant of K-fold to maintain the ratio of healthy vs. non-healthy. 
I used the random permutation cross validation scheme because it results in more splits (if set so), thus more reliable score averages. 

2. Good point. With score I meant the F1 score, as was used in the publication that was provided with the assignment. I also tried the ROC AUC value, but I ran into a practical problem with Python package I used. 

Concerning your last comment: the stability between folds is also something worth looking at as a measure of performance. I have not looked at it that way, so thank you.

**Comment by sreejita-rug** [comment]
F1 score is also fine. ROC is not a must. Any measure that takes the imbalance into account is good. You can also look at Matthew correlation coefficient, if you wish you and if that can be easily computed by any of the packages you are using in Python. 
In the end for a particular classifier with a particular setting, check the average performance over the N fold for training and validation, in order to decide which is a better classifier. If you look at figure 1 of this Assignment,  you'll get some hints about how to report your findings. What we meant by the top 5 classifier in "Remember to tabularize the top 5 classifier experimental settings in your report" is this: 
1) Let's call one of your best performing classifier Merlin.  Merliln is actually one of the classifiers (say KNN) with a particular strategy to deal with some of the issues this real world medical data poses (let's call this this is Experiment I), and in this experiment you try different settings to check if the classifier performance (models in all the N fold have similar good class-wise accuracy, there is less overfitting in all the folds) has improved. Now when you are satisfied with your Merlin which is KNN in Experiment I with Setting S1, you can have Merlin as one of the candidates for your ensemble and also report this detail in table of the top 5 classifiers. While reporting, report the average performance of Merlin over the N folds (the standard deviation values will give you an idea about the stability of the models over the N folds).
2) Next you have suppose this other classifier, Gandalf, which is actually Decision Tree with a certain amount of pruning, and certain strategies to handle the high dimensionality or other issues of this real world medical data (let's call this Experiment J). And again here you had tried different settings in your Experiment J (say varying a parameter in your experimental setup, from 200% to 400%, or decrease another parameter value from 0.05 to 0.01 or 0.001). And for a certain setting in Experiment J the Decision Tree gave a wonderful performance (what I mean by wonderful performance is already described above in (1) in my previous comments), then this combination of Decision Tree and the experiment J, and setting S5, is Gandalf -the other candidate for your ensemble and another of the top 5 models you are to tabularize, along with the performance measures like classwise accuracy, F1 score or MCC or AUC from ROC, sensitivity, specificity, etc.
3) Now suppose again with KNN and with Experiment J, and setting S3 you find disappointing results (model unstable over N folds, or overfitting, or model is too simple, whatever!). And let's call this combination of KNN and experiment J and setting S3, Rincewind. Since Rincewind disappointed you it is advisable not to include Rincewind in the ensemble and ofcourse not mention it in the table of the top 5. 

And so on... In the report we expect you to describe briefly the Experiments you performed, along with the pictorial summary (an example of which is given in figure 1 of assignment 1). Hope everything is clear now. In case something isn't clear just ask now and before tomorrow noon rather than asking on the 11th hour and freaking out.  A few wise men and women have said, "there is no such thing as stupid question". 
Good luck! :)

**Comment by jasperdev** [comment]
Ok sounds good. Thanks for the help!
